### ISSUE TITLE
Sending IP

### ISSUE THREAD

**Issue by RobinFlikkema** [mailcow/mailcow-dockerized#302]
How do i change the IP which get used to send email? My server has multiple (virtual) interfaces and by default it uses eth0 but i want to use eth0:3 for sending email (to other mailservers, like gmail)

By changing the mailcow.conf i managed to set the correct IP for most services (like IMAP, HTTP(S) etc)

This is a debian 8 machine using mailcow-dockerized

**Comment by elvirdz** [comment]
bind ip on eth0:3 not on 0.0.0.0

**Comment by RobinFlikkema** [comment]
But i need to change the IP that postfix uses right? And postfix doesn't accept eth0:3 (probably because it is in a container)
`May 20 09:27:05 mail postfix[229]: fatal: config variable inet_interfaces: host not found: eth0:3`

**Comment by elvirdz** [comment]
you can not bind eth0:3 you can only bind on IP from eth0:3 
i hope u have for every ethport other ipv4

**Comment by RobinFlikkema** [comment]
Well, I also can't bind to an IP from within the postfix config. I think this is due to the fact that docker uses its own networking

Postfix main.cf:
>inet_interfaces = 5.xxx.xxx.251

Postfix container log:
>May 20 09:45:36 mail postfix[229]: fatal: parameter inet_interfaces: no local interface found for 5.xxx.xxx.251

/etc/network/interfaces:

> auto eth0:3    
>iface eth0:3 inet static    
>    address 5.xxx.xxx.251    
>    netmask 255.255.255.255    
>    broadcast 5.xxx.xxx.251

(Where the x'es are obviously replaces by the real octets.)

**Comment by elvirdz** [comment]
dont know u have to wait on andryyy

i think you can not do that ... container

**Comment by andryyy** [comment]
You can set a SNAT ip for the mailcow network like =>

```
iptables -t nat -I POSTROUTING -p all -s 172.22.1.0/24 -j SNAT --to-source 1.2.3.4
```

**Comment by RobinFlikkema** [comment]
Okay, so when I do _nothing_, it works but sends from 51.xxx.xxx.183
> 2017-5-20 16:26:46	info	8404C23185: to=, relay=gmail-smtp-in.l.google.com[74.125.133.26]:25, delay=3.5, delays=0.22/0/3.1/0.19, dsn=2.0.0, status=sent (250 2.0.0 OK 1495290406 k48si6846388wrk.201 - gsmtp)
>2017-5-20 16:26:45	info	Trusted TLS connection established to gmail-smtp-in.l.google.com[74.125.133.26]:25: TLSv1.2 with cipher ECDHE-RSA-CHACHA20-POLY1305 (256/256 bits)

When i use your iptables-snippet (where i replaced 1.2.3.4 by 5.xxx.xxxx.251) it doesnt work:
> 2017-5-20 16:30:45	info	disconnect from unknown[5.xxxx.xxx.251] ehlo=1 auth=1 mail=1 rcpt=1 data=1 quit=1 commands=6
> 2017-5-20 16:30:45	info	CC9E723185: from=, size=1056, nrcpt=1 (queue active)
> 2017-5-20 16:30:45	info	CC9E723185: message-id=<24-59205300-1-22efa580@150578145>
> 2017-5-20 16:30:45	info	CC9E723185: client=unknown[5.xxx.xxx.251], sasl_method=PLAIN, sasl_username=robin@xxxxxx.nl
> 2017-5-20 16:30:45	info	connect from unknown[5.xxx.xxx.251]
> 2017-5-20 16:30:36	info	D6E8D23186: to=, relay=none, delay=573, delays=573/0.05/0/0, dsn=4.4.3, status=deferred (Host or domain name not found. Name service error for name=gmail.com type=MX: Host not found, try again)
> 2017-5-20 16:30:36	info	D6E8D23186: from=, size=1065, nrcpt=1 (queue active)

It seems like when I use SNAT it can't resolve gmail's MX records or something.
To be sure, i restarted all mailcow docker containers and tried again, but it gives the same error:
> D6E8D23186: to=, relay=none, delay=1284, delays=1284/0.01/0/0, dsn=4.4.3, status=deferred (Host or domain name not found. Name service error for name=gmail.com type=MX: Host not found, try again)
Although I didn't see a reason why it was needed, I also tried adding 5.xxx.xxx.251 to postfix/main.cf but:
> 	fatal: parameter inet_interfaces: no local interface found for 5.xxx.xxx.251

The interface itselfs works fine AFAIK:
> root@giraffe:~# ping -I 5.xxx.xxx.251 google.nl
> PING google.nl (216.58.209.227) from 5.xxx.xxx.251 : 56(84) bytes of data.
> 64 bytes from par10s29-in-f227.1e100.net (216.58.209.227): icmp_seq=1 ttl=53 time=4.68 ms
> 64 bytes from par10s29-in-f227.1e100.net (216.58.209.227): icmp_seq=2 ttl=53 time=4.57 ms
> 64 bytes from par10s29-in-f227.1e100.net (216.58.209.227): icmp_seq=3 ttl=53 time=4.60 ms
> ^C
> --- google.nl ping statistics ---
> 3 packets transmitted, 3 received, 0% packet loss, time 2003ms
> rtt min/avg/max/mdev = 4.573/4.622/4.684/0.046 ms

> root@giraffe:~# ifconfig
> (other interfaces)
> eth0:3: flags=4163<UP,BROADCAST,RUNNING,MULTICAST>  mtu 1500
>        inet 5.xxx.xxx.251  netmask 255.255.255.255  broadcast 5.xxx.xxx.251
>        ether fa:16:3e:7e:a1:f8  txqueuelen 1000  (Ethernet)

**Comment by andryyy** [comment]
The postrouting should work fine, but I don't know your network. :/

Maybe it is the reverse-path filter to block it? See "net.ipv4.conf.all.rp_filter" and "net.ipv4.conf.default.rp_filter".

**Comment by RobinFlikkema** [comment]
> Maybe it is the reverse-path filter to block it? See "net.ipv4.conf.all.rp_filter" and "net.ipv4.conf.default.rp_filter"

These were commented out in sysctl.conf so they should default to 0/off. I removed the comment and made sure they were set to 0. And confirmed changes with `sysctl -p`

> The postrouting should work fine, but I don't know your network. :/

I use "legacy method" to assign multiple virtual interfaces. (see [Debian wiki](https://wiki.debian.org/NetworkConfiguration#Legacy_method))

> root@giraffe:~# cat /etc/network/interfaces
> \# The loopback network interface
> auto lo
> iface lo inet loopback
> 
> auto eth0
> iface eth0 inet static
>         address 51.xxx.xxx.183
>         netmask 255.255.255.255
>         post-up /sbin/ip route add 51.xxx.xxx.1 dev eth0
>         post-up /sbin/ip route add default via 51.xxx.xxx.1
>         pre-down /sbin/ip route del default via 51.xxx.xxx.1
>         pre-down /sbin/ip route del 51.xxx.xxx.1 dev eth0
>         dns-nameserver 213.186.33.99
>         dns-search ovh.net
> iface eth0 inet6 static
>         address 2001:41d0:xxx:xxxx::43b5
>         netmask 128
>         post-up /sbin/ip -6 route add 2001:41d0:xxx:xxxx::1 dev eth0
>         post-up /sbin/ip -6 route add default via 2001:41d0:xxx:xxxx::1 dev eth0
>         pre-down /sbin/ip -6 route del default via 2001:41d0:xxx:xxxx::1 dev eth0
>         pre-down /sbin/ip -6 route del 2001:41d0:xxx:xxxx::1 dev eth0
> 
> auto eth0:0
> iface eth0:0 inet static
>         address 193.xx.xx.212
>         netmask 255.255.255.255
>         broadcast 193.xx.xx.212
> 
> auto eth0:1
> iface eth0:1 inet static
>         address 193.xx.xxx.128
>         netmask 255.255.255.255
>         broadcast 192.xx.xxx.128
> 
> auto eth0:2
> iface eth0:2 inet static
>         address 193.xx.xxx.129
>         netmask 255.255.255.255
>         broadcast 192.xx.xxx.129
> 
> auto eth0:3
> iface eth0:3 inet static
>         address 5.xxx.xxx.251
>         netmask 255.255.255.255
>         broadcast 5.xxx.xxx.251
> (more interfaces)

This server is hosted at OVH France, and my server uses their DNS cache as primary resolver.

The Mailcow network hasn't changed and is still default. And thus uses 172.22.1.0/24.

But don't you think this is DNS related? It seems to me that Postfix can't reach a DNS server after using POSTROUTE. Maybe because the reply doesn't get routed properly to postfix?
![Imgur](http://i.imgur.com/QkmhFZV.png?1)

**Comment by RobinFlikkema** [comment]
It's definitely the routing of DNS or something. When using `ping` from within the postfix container you can see that it can't resolve domainnames.
> root@giraffe:~# docker exec -it mailcowdockerized_postfix-mailcow_1 /bin/bash
> root@mail:/# ping 8.8.8.8 -c 4
> PING 8.8.8.8 (8.8.8.8): 56 data bytes
> 64 bytes from 8.8.8.8: icmp_seq=0 ttl=55 time=5.059 ms
> 64 bytes from 8.8.8.8: icmp_seq=1 ttl=55 time=4.882 ms
> 64 bytes from 8.8.8.8: icmp_seq=2 ttl=55 time=4.994 ms
> 64 bytes from 8.8.8.8: icmp_seq=3 ttl=55 time=4.895 ms
> --- 8.8.8.8 ping statistics ---
> 4 packets transmitted, 4 packets received, 0% packet loss
> round-trip min/avg/max/stddev = 4.882/4.957/5.059/0.073 ms
> root@mail:/# ping google.nl -c 4
> ping: unknown host

But this works fine when not using SNAT:
> root@giraffe:~# docker exec -it mailcowdockerized_postfix-mailcow_1 /bin/bash
> root@mail:/# ping 8.8.8.8 -c 4
> PING 8.8.8.8 (8.8.8.8): 56 data bytes
> 64 bytes from 8.8.8.8: icmp_seq=0 ttl=55 time=4.957 ms
> 64 bytes from 8.8.8.8: icmp_seq=1 ttl=55 time=4.920 ms
> 64 bytes from 8.8.8.8: icmp_seq=2 ttl=55 time=4.875 ms
> 64 bytes from 8.8.8.8: icmp_seq=3 ttl=55 time=4.869 ms
> --- 8.8.8.8 ping statistics ---
> 4 packets transmitted, 4 packets received, 0% packet loss
> round-trip min/avg/max/stddev = 4.869/4.905/4.957/0.036 ms
> root@mail:/# ping google.nl -c 4
> PING google.nl (216.58.206.227): 56 data bytes
> 64 bytes from 216.58.206.227: icmp_seq=0 ttl=52 time=4.593 ms
> 64 bytes from 216.58.206.227: icmp_seq=1 ttl=52 time=4.658 ms
> 64 bytes from 216.58.206.227: icmp_seq=2 ttl=52 time=4.633 ms
> 64 bytes from 216.58.206.227: icmp_seq=3 ttl=52 time=4.703 ms
> --- google.nl ping statistics ---
> 4 packets transmitted, 4 packets received, 0% packet loss
> round-trip min/avg/max/stddev = 4.593/4.647/4.703/0.040 ms

**Comment by andryyy** [comment]
But you could have tested it by pining an IP address. Not a name. 

Ovh also has these many restrictions for routing IPs. I'm sure you gave them a MAC address. 

Furthermore OVH is on many blacklists. 

Am 21.05.2017 um 10:00 schrieb Robin Flikkema <notifications@github.com>:

**Comment by lenovouser** [comment]
I have the same problem - the DNS resolving doesn't work anymore:

```sh
curl: (6) Could not resolve host: icanhazip.com
```

It doesn't seem to be a routing problem though. Pinging works and when I bypass the DNS with curl I also get the correct IP back from [canihazip.com](http://canihazip.com):

```sh
curl --verbose --header 'Host: canihazip.com' 'http://5.196.192.222:80/s'
* Trying 5.196.192.222...
* TCP_NODELAY set
* Connected to 5.196.192.222 (5.196.192.222) port 80 (#0)
> GET /s HTTP/1.1
> Host: canihazip.com
> User-Agent: curl/7.52.1
> Accept: */*
>
< HTTP/1.1 200 OK
< Date: Mon, 22 May 2017 17:24:54 GMT
< Server: Apache
< X-Powered-By: PHP/5.5.9-1ubuntu4.14
< Access-Control-Allow-Origin: *
< X-Content-Type-Options: nosniff
< X-Frame-Options: sameorigin
< Content-Length: 15
< Content-Type: text/plain
<
* Curl_http_done: called premature == 0
* Connection #0 to host 5.196.192.222 left intact
<my ip>
```

**Comment by RobinFlikkema** [comment]
While i tried adding my IP as a allowed IP in de bind9 container, could it be that iptables is changing the IP of the DNS lookups, and bind9 is refusing the query because it only allows RFC1918 IPs?

**Comment by RobinFlikkema** [comment]
I moved the mailserver to another server so there is no reason for me to change the source IP anymore.

**Comment by Ringelnatz** [comment]
After hours of research I found out that the SNAT postrouting is working, if the "dns:" section is removed from the postfix container in docker-compose.yml. Mails now get sent as requested which takes much more time now (because of the removal of the dns section?).

What would be a more elegant way to get name resolving back working with the SNAT rule enabled?

**Comment by Ringelnatz** [comment]
I finally fixed this by allowing my desired outgoing ip address in `unbound.conf`.

Because this is a problem affecting every host with more than one ip address attached to it, I think there maybe should be a way to configure this properly.

**Comment by Ringelnatz** [comment]
I for now ended up with these two steps in order to specify an ip address for sending emails (because the iptables command has to be run after every container (re)start):

1. edit `data/conf/unbound/unbound.conf` and add `access-control: MY-IP-ADDRESS allow` (i guess this has to be done after every mailcow update)
2. run the following script at boot time (as root):

```bash
#!/bin/bash

sleep 10

if [[ "$(docker inspect -f {{.State.Running}} mailcow-dockerized_postfix-mailcow_1)" == "true" ]]; then
        /sbin/iptables -t nat -I POSTROUTING -s 172.22.1.0/24 -j SNAT --to-source MY-IP-ADDRESS
fi


docker events --filter 'event=start' --filter 'event=stop' --filter 'container=mailcow-dockerized_postfix-mailcow_1' | while read event

do

        if [[ $event == *"container start"* ]]; then
                /sbin/iptables -t nat -I POSTROUTING -s 172.22.1.0/24 -j SNAT --to-source MY-IP-ADDRESS
        else
                /sbin/iptables -t nat -D POSTROUTING -s 172.22.1.0/24 -j SNAT --to-source MY-IP-ADDRESS
        fi

done
```

Hopefully docker will be able to do step 2 on it's own one day (e.g. `docker run ... --outbound-ip 1.2.3.4`).

**Comment by stAn47** [comment]
hello friends, even though this seems to be closed, i am not sure if routing is required to rewrite the "sender IP" address. 

my problem is that all outgoing emails send via postfix which are authenticated via smtp include this header: 

Received: from MYDOMAIN.com ([172.22.1.9])

which is obviously a private address. 

it seems that one of the solutions since MYDOMAIN.com is the only outgoing email domain (mx) is to modify the /etc/hosts file within postfix docker image

i commented this line: 
```
#172.22.1.9 MYDOMAIN.com mx
```
and added 2 new lines:
```
{public ipv4} MYDOMAIN.com mx
{public ipv6} MYDOMAIN.com mx
```
now when i send emails, the email header properly says the IP address to which the domain resolves 

can anybody confirm that my solution is OK ? 

i understand that host file will need to be updated after each maicow update. 

to edit the hosts file i've used: 
docker exec -it mailcowdockerized_postfix-mailcow_1 /bin/bash
to enter bash and apt-get install to get editor installed (mc)

thank you, best regards, stan
